### PR TITLE
add teams connector

### PR DESCRIPTION
**Pull request by refarer**

### CODE REVIEW

File: doc/TeamsConnector.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+[Code](https://github.com/reshufflehq/reshuffle-microsoft-connectors) |
+[npm](https://www.npmjs.com/package/reshuffle-microsoft-connectors) |
+[Code sample](https://github.com/reshufflehq/reshuffle/tree/master/examples/microsoft/teams)
+
+`npm install reshuffle-microsoft-connectors`
+
+### Reshuffle Microsoft Connectors
+
+This package contains a collection of [Reshuffle](https://github.com/reshufflehq/reshuffle)
+connectors to [Microsoft](https://microsoft.com).
```
Comment (foundingnimo) [inline doc/TeamsConnector.md]: I think for the connector-specific docs, we should follow the AWS connectors' pattern where we use the term connector. E.g. 
`This Reshuffle connector can be used to access AWS Lambda. It is implemented using Amazon's Lambda SDK.`

I know the Excel readme does not follow this pattern. I will submit a PR

File: doc/TeamsConnector.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+The following example adds a new worksheet to an excel file
```
Comment (foundingnimo) [inline doc/TeamsConnector.md:9]: Copy/pasta?

File: doc/TeamsConnector.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+```js
+const { Reshuffle } = require('reshuffle')
+const { TeamsConnector } = require('reshuffle-microsoft-connectors')
+
+const app = new Reshuffle()
+const connector = new TeamsConnector(app, { AppId, AppPassword, AppTenantId })
```
Comment (foundingnimo) [inline doc/TeamsConnector.md]: I think we should keep the convention of `process.env` for arbitrary parameters, WDYT?

File: doc/TeamsConnector.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+[listChannels](#listChannels) List Channels
+
+##### Configuration options
+
+```js
+const app = new Reshuffle()
+const connector = new TeamsConnector(app, { AppId, AppPassword, AppTenantId })
+```
+
+Credentials can be created by following the guide at https://docs.microsoft.com/en-us/graph/auth-v2-service
```
Comment (foundingnimo) [inline doc/TeamsConnector.md:50]: Could we link the guide please?

File: doc/TeamsConnector.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    notificationUrl: 'https://example.com/webhook',
+    expirationDateTime: '2021-02-03T03:47:17.292Z',
+  },
+  (event) => console.log(event),
+)
+
+connector.listTeams().then((teams) => console.log(teams))
+```
+
+#### Table of Contents
```
Comment (foundingnimo) [inline doc/TeamsConnector.md:31]: Can we also list events please?

File: src/base.ts [hunk 6]
```diff
@@ -1,12 +1,19 @@
...
+import type MicrosoftGraph from '@microsoft/microsoft-graph-types'
+import type { Request, Response } from 'express'
 
 import { ClientCredentialAuthenticationProvider, MicrosoftOptions } from './auth'
 
 export type MicrosoftConnectorConfigOptions = MicrosoftOptions & { debugLogging?: boolean }
+
+export type MicrosoftEventConfigOptions = MicrosoftGraph.Subscription
+
+const DEFAULT_WEBHOOK_PATH = '/reshuffle-microsoft-connector/webhook'
```
Comment (foundingnimo) [inline src/base.ts]: In both Monday and quickbooks we added a convention of `/webhooks/xxx` where `xxx` is the service name. I think we should stick to this?

File: src/base.ts [hunk 7]
```diff
@@ -21,6 +28,85 @@ export class MicrosoftConnector extends BaseHttpConnector<
     this.client = Client.initWithMiddleware(clientOptions)
   }
 
+  async onStart(): Promise<void> {
+    const logger = this.app.getLogger()
+    const subscriptions: MicrosoftGraph.Subscription[] = (
+      await this.sdk().api('/subscriptions').get()
```
Comment (foundingnimo) [inline src/base.ts:38]: This should be in `try{....} catch(...`

File: src/base.ts [hunk 8]
```diff
@@ -21,6 +28,85 @@ export class MicrosoftConnector extends BaseHttpConnector<
...
+        ),
+      )
+      let subscription: MicrosoftGraph.Subscription
+      if (existingSubscription) {
+        logger.info(
+          `Reshuffle Microsoft - existing webhook reused (resource: ${existingSubscription.resource}, url: ${existingSubscription.notificationUrl})`,
+        )
+        subscription = existingSubscription
+      } else {
+        subscription = await this.sdk().api('/subscriptions').post(options)
```
Comment (foundingnimo) [inline src/base.ts]: This should also be in `try{... catch(...}`

File: src/base.ts [hunk 9]
```diff
@@ -21,6 +28,85 @@ export class MicrosoftConnector extends BaseHttpConnector<
...
+  async handle(req: Request, res: Response): Promise<boolean> {
+    const { validationToken } = req.query
+
+    if (validationToken) {
+      res.send(validationToken)
+    } else {
+      const data = req.body.value
+
+      for (const incomingEvent of data) {
+        const eventsUsingGithubEvent = Object.values(this.eventConfigurations).filter(
```
Comment (foundingnimo) [inline src/base.ts]: Github?

File: src/base.ts [hunk 10]
```diff
@@ -21,6 +28,85 @@ export class MicrosoftConnector extends BaseHttpConnector<
...
+          (event: EventConfiguration) =>
+            event.options.subscriptionId === incomingEvent.subsciptionId,
```
Comment (foundingnimo) [inline src/base.ts]: is `subsciptionId` typo in the data coming in?

File: src/base.ts [hunk 11]
```diff
@@ -21,6 +28,85 @@ export class MicrosoftConnector extends BaseHttpConnector<
...
+
+        for (const event of eventsUsingGithubEvent) {
+          await this.app.handleEvent(event.id, {
+            ...event,
+            ...req.body,
+          })
+        }
+      }
+
+      res.status(202).send()
```
Comment (foundingnimo) [inline src/base.ts]: Should this be sent before the `await`ing loop? How fast is Microsoft expecting the response to return?

File: src/base.ts [hunk 12]
```diff
@@ -86,14 +95,12 @@ export class MicrosoftConnector extends BaseHttpConnector<
...
       res.send(validationToken)
     } else {
-      const data = req.body.value
-
+      const { value: data = [] } = req.body as MicrosoftGraph.ChangeNotificationCollection
       for (const incomingEvent of data) {
         const eventsUsingGithubEvent = Object.values(this.eventConfigurations).filter(
           (event: EventConfiguration) =>
-            event.options.subscriptionId === incomingEvent.subsciptionId,
+            event.options.subscriptionId === incomingEvent.subscriptionId,
```
Comment (foundingnimo) [inline src/base.ts:104]: :)

File: src/base.ts [hunk 13]
```diff
@@ -35,10 +38,16 @@ export class MicrosoftConnector extends BaseHttpConnector<
     ).value
 
     for (const event of Object.values(this.eventConfigurations)) {
-      const options: MicrosoftEventConfigOptions = event.options
+      const { runtimeBaseUrl, ...options } = event.options as MicrosoftConnectorConfigOptions
+      const eventSubscription = {
+        ...options,
+        notificationUrl: runtimeBaseUrl + DEFAULT_WEBHOOK_PATH,
```
Comment (foundingnimo) [inline src/base.ts]: Is it the DEFAULT_WEBHOOK_PATH here, or do we let the user alter the value?

**Review by foundingnimo — CHANGES_REQUESTED** [review]

File: src/base.ts (hunk 13 above)
Comment (refarer) [reply src/base.ts]: I'll add it as an option

File: src/base.ts (hunk 11 above)
Comment (refarer) [reply src/base.ts]: You have 30 seconds to respond. I think you only want to send a success response once the events have all been handled.

Comment (foundingnimo) [reply src/base.ts]: But this is not a success response. It is 'accepted'. I think the meaning of it is exactly "I received your request and am now processing it"

Comment (refarer) [reply src/base.ts]: Yeah your right. I'll move it to before the requests are handled.

**Review by foundingnimo — APPROVED** [review]
Cheers!

**Pull request merged**